Modules/xASL_module_Import.m

```diff
@@ -105,123 +113,115 @@
     end
 
-    % By default don't copy DICOMs for anonymization reasons
-    if nargin<5 || isempty(bCopySingleDicoms)
-        bCopySingleDicoms = false;
+    
+    %% 2. Run the NIIX to ASL-BIDS
+    iState = 2;
+    if x.opts.ImportModules(2) && ~x.mutex.HasState(StateName{2})
+        x = xASL_imp_NII2BIDS(x, imPar);
+        x.mutex.AddState(StateName{iState});
+    elseif x.opts.ImportModules(2) && x.mutex.HasState(StateName{2})
+        fprintf('NIIX to ASL-BIDS was run before...   \n');
     end
 
-    if nargin<6 || isempty(bUseDCMTK)
-        % Default set to using DCM-TK
-        bUseDCMTK = true;
-    elseif ~bUseDCMTK && isempty(which('dicomdict'))
-        error('Dicomdict missing, image processing probably not installed, try DCMTK instead');
+    
+    %% 3. Run defacing
+    iState = 3;
+    if x.opts.ImportModules(3) && ~x.mutex.HasState(StateName{3})
+        xASL_imp_Deface(x,imPar);
+        x.mutex.AddState(StateName{iState});
+    elseif x.opts.ImportModules(3) && x.mutex.HasState(StateName{3})
+        fprintf('Defacing was run before...   \n');
     end
-
-    if nargin<7 || isempty(bCheckPermissions)
-        if isunix
-            bCheckPermissions = true;
-        else
-            bCheckPermissions = false;
-        end
+    
+    
+    %% 4. Run BIDS to Legacy
+    iState = 4;
+    if x.opts.ImportModules(4) && ~x.mutex.HasState(StateName{4})
+        x = xASL_imp_BIDS2Legacy(x);
+        x.mutex.AddState(StateName{iState});
+    elseif x.opts.ImportModules(4) && x.mutex.HasState(StateName{4})
+        fprintf('BIDS to Legacy was run before...   \n');
     end
 
-    % Only initialize ExploreASL if this wasn't initialized before
-    if nargin<8 || isempty(x)
-        x = ExploreASL_Initialize;
-    end
     
-    % Write import settings to modules field of x structure
-    x.modules.import.settings.bCopySingleDicoms = bCopySingleDicoms;
-    x.modules.import.settings.bUseDCMTK = bUseDCMTK;
-    x.modules.import.settings.bCheckPermissions = bCheckPermissions;
+    %% 5. Clean-up
+    [x] = xASL_imp_CleanUpImport(x);
+    result = x.result;
 
-    %% 2. Initialize the import setup
     
-    % Load the sourceStructure.json and initialize the corresponding struct
-    if bRunSubmodules(1) || bRunSubmodules(2) || bRunSubmodules(3)
-        imPar = xASL_imp_Initialize(studyPath, imParPath);
-    end
+end
 
-    %% 3. Run the DCM2NIIX
-    if bRunSubmodules(1)
-        xASL_imp_DCM2NII(imPar, x);
-    end
 
-    %% 4. Run the NIIX to ASL-BIDS
-    if bRunSubmodules(2)
-        % Run NII to BIDS
-        x = xASL_imp_NII2BIDS(x, imPar, studyPath, studyParPath);
-        % Update x.opts.DatasetRoot
-        x = xASL_imp_UpdateDatasetRoot(x, studyPath);
-    end
+%% Clean-Up before processing pipeline
+function [x] = xASL_imp_CleanUpImport(x)
 
-    %% 5. Run defacing
-    if bRunSubmodules(3)
-        xASL_imp_Deface(imPar);
+    % We want to reload the derivatives data correctly, which is why we delete the following 
```

Comment:
Are you sure these are all the overlapping fields? E.g., `x.SUBJECTS` should also be removed for import, and some more right?

Reply:
And that's happening in the clean up? Afterwards we definitely need it, otherwise you can't run any import with xASL_Iteration.

Reply:
The problem here — or perhaps not here but what I meant — is that `x.SUBJECTS` is the same for the structural, ASL, and population modules, but may differ for the import module. So for now we should definitely remove it in the end of the import module (I don't know what happens in the clean up :) and it needs to be recreated for the processing modules.

What would be nicer, is that each module separately fills `x.SUBJECTS`, rather than taking it from the general initialization. So we would split the initialization into at least 2 functions:

1) initialisation of ExploreASL
2) loading of data/defining subjects/sessions/runs: this is reiterated before each module

What do you think? In a new issue of course.

Reply:
> What would be nicer, is that each module separately fills `x.SUBJECTS`, rather than taking it from the general initialization. 

This doesn't work since it is a part of xASL_Iteration and you probably won't be able to get rid of it there.